### PR TITLE
Get PMC MSR bit width from CPUID

### LINKED ISSUE #176: Prototype setting programmable counters in independent test

### LINKED ISSUE #1931: MSR::IA32_PMC0:PERFCTR is not always the right width
**Describe the bug**
I tried to compare the fixed-counter and programmable-counter variants of instructions retired. The fixed-counter variant produced consistent differences across a sleep, but the programmable counter often reported negative differences. This happens because GEOPM discards upper bits that are in use for some architectures.

**Possible Fix**
SDM vol 3b, section 18 specifies that we should read `CPUID.0AH:EAX[23:16]` to learn how many bits are available for programmable counters (48 on our skylake nodes). service/src/msr_data_arch.cpp is currently configured to read 31 bits for `MSR::IA32_PMC0:PERFCTR`. 

MSRIOGroup.cpp already has a workflow for MSRs with bit width that depends on CPUID. See the handlers for `QM_EVTSEL:RMID` and `PQR_ASSOC:RMID` for reference.

**Steps to reproduce**
```
# Read the fixed counter for retired instructions.
for i in {1..5}; do i1=$(geopmread INSTRUCTIONS_RETIRED board 0) && sleep 2 && i2=$(geopmread INSTRUCTIONS_RETIRED board 0) && echo "FIXED COUNTER Instructions: $((i2-i1))"; done
```
```
# Enable INST_RETIRED.ANY_P. See https://download.01.org/perfmon/SKX/skylakex_core_v1.24.json
geopmwrite MSR::IA32_PERFEVTSEL0:EVENT_SELECT board 0 0xc0
geopmwrite MSR::IA32_PERFEVTSEL0:UMASK board 0 0x00
geopmwrite MSR::IA32_PERFEVTSEL0:USR board 0 1
geopmwrite MSR::IA32_PERFEVTSEL0:OS board 0 1
geopmwrite MSR::IA32_PERFEVTSEL0:EN board 0 1
geopmwrite  MSR::PERF_GLOBAL_CTRL:EN_PMC0 board 0 1
# Read the PMC for retired instructions. Should match the measurements from before.
for i in {1..5}; do i1=$(geopmread MSR::IA32_PMC0:PERFCTR board 0) && sleep 2 && i2=$(geopmread MSR::IA32_PMC0:PERFCTR board 0) && echo "Programmable Counter Instructions: $((i2-i1))"; done
```

**GEOPM version**
1.1.0+dev1567gc9a848c8

### PR DESCRIPTION
**Pull request by dannosliwcd**
- Resolves #176 by adding an integration test for the instructions retired programmable counter.
- Fixes #1931 by using width of MSR::IA32_PMC0:PERFCT that is reported by CPUID.

### CODE REVIEW

File: service/src/msr_data_arch.cpp [hunk 1]
```diff
@@ -814,7 +814,7 @@ namespace geopm
             "fields": {
                 "PERFCTR": {
                     "begin_bit": 0,
-                    "end_bit":   31,
+                    "end_bit":   47,
```
Comment (lhlawson) [inline service/src/msr_data_arch.cpp:817]: Is there any way to communicate to the end-user that this value won't be used?

File: service/src/MSRIOGroup.cpp [hunk 2]
```diff
@@ -1320,6 +1336,12 @@ namespace geopm
                     }
                 }
 
+                if (string_begins_with(msr_field_name, "IA32_PMC") &&
+                    string_ends_with(msr_field_name, ":PERFCTR") &&
+                    m_pmc_bit_width > 0) {
+                    end_bit = begin_bit + m_pmc_bit_width - 1;
+                }
```
Comment (lhlawson) [inline service/src/MSRIOGroup.cpp]: Should we emit a specific warning in debug mode if m_pmc_bit_wdith == 0 to indicate we're using the JSON value?

**Review by lhlawson — COMMENTED** [review]
LGTM overall, 2 specific questions on communication to the user.

File: service/src/msr_data_arch.cpp (hunk 1 above)
Comment (cmcantalupo) [reply service/src/msr_data_arch.cpp:817]: You can't comment in json, and the "end-user" shouldn't really be poking around in this data.  We could put a comment about this fact near the code that overwrites the value?

**Review by cmcantalupo — COMMENTED** [review]
LGTM.  Seems like @lhlawson has some feedback which can be addressed.

File: service/src/MSRIOGroup.cpp (hunk 2 above)
Comment (dannosliwcd) [reply service/src/MSRIOGroup.cpp]: Good call, will do. That's what was causing the GitHub Actions failure, by the way (CPUID was saying we get 0 bits, and we don't support 0-bit MSRs)

File: service/src/MSRIOGroup.cpp [hunk 3]
```diff
@@ -1320,6 +1336,19 @@ namespace geopm
...
+#ifdef GEOPM_DEBUG
+                else {
+                    std::cerr << "Warning: <geopm> CPUID specified 0 bits for "
+                              << msr_field_name << "; using the default width: "
+                              << (end_bit - begin_bit + 1) << std::endl;
```
Comment (dannosliwcd) [inline service/src/MSRIOGroup.cpp]: This needs to be inside the begins/endswith check.

**Review by cmcantalupo — DISMISSED** [review]
LGTM

**Review by cmcantalupo — APPROVED** [review]
Thanks!

**Pull request merged**